### PR TITLE
[Java] Support actor handle reference counting.

### LINKED ISSUE #10976: Support actor handle ref counting for Java
Corresponding Python PR: https://github.com/ray-project/ray/pull/7434

### PR DESCRIPTION
**Pull request by jovany-wang**
## Related issue number
Close #10976

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jovany-wang** [comment]
This PR is finished, but it's blocked by #21074 . Will request reviews after #21074 merged.

File: java/test/src/main/java/io/ray/test/ActorHandleReferenceCountTest.java
Comment (jovany-wang) [inline java/test/src/main/java/io/ray/test/ActorHandleReferenceCountTest.java:88]: This case is blocked at this PR https://github.com/ray-project/ray/pull/21279

Comment (jovany-wang) [reply java/test/src/main/java/io/ray/test/ActorHandleReferenceCountTest.java:88]: In our current master codebase, there is no approach to get `num_java_worker_per_process` config. I'm aiming to address it at #21279 .

**Comment by jovany-wang** [comment]
Hi @kfstorm @raulchen , this is ready for review.

File: src/ray/gcs/gcs_server/gcs_job_manager.cc
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:128]: Is this not needed?

File: src/ray/gcs/gcs_server/gcs_job_manager.cc
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_job_manager.cc:179]: Nit, I think we only need to expose a `GetJobConfig` method, and use `GetJobConfig().num_java_workers_per_process()` at the call sites. 
Otherwise there will be too many such methods.

**Review by raulchen — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_job_manager.cc
Comment (kfstorm) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:179]: +1

**Review by kfstorm — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_job_manager.cc
Comment (jovany-wang) [reply src/ray/gcs/gcs_server/gcs_job_manager.cc:179]: Good point. I'm going to refine this at #21279 because there still be an issue that when we should clear the cache.

**Comment by jovany-wang** [comment]
tests:test_client failed on master.

**Pull request merged**
